### PR TITLE
Added opening streams and outputting to a stream

### PR DESCRIPTION
**Pull request by visoft**
This PR adds the ability to open streams instead of just local files. This work was done before by @tokichie, but that branch was deleted.

This adds opening a stream and outputting to a stream (useful for HTTP transfers).

### CODE REVIEW

**Comment by mvz** [comment]
@satoryu any chance of getting this merged (minus the version bump)?

File: lib/docx/document.rb [hunk 1]
```diff
@@ -20,9 +20,11 @@ module Docx
   class Document
     attr_reader :xml, :doc, :zip, :styles
 
-    def initialize(path, &block)
+    def initialize(path_or_io, options = {})
       @replace = {}
-      @zip = Zip::File.open(path)
+
+      @zip = options[:use_buffer] ? Zip::File.open_buffer(path_or_io) : Zip::File.open(path_or_io)
```
Comment (satoryu) [inline lib/docx/document.rb]: @visoft your idea looks good. but users can pass `use_buffer: true` with a file path. it should be prevented if possible. 
This means that `use_buffer` option can be removed if we have a way to detect whether `path_or_io` is a `IO` or a `String`.

File: lib/docx/version.rb [hunk 2]
```diff
@@ -1,3 +1,3 @@
-module Docx #:nodoc:
-  VERSION = '0.3.0'
-end
+module Docx #:nodoc:
+  VERSION = '0.3.1'
```
Comment (satoryu) [inline lib/docx/version.rb]: please revert this update.

**Comment by mvz** [comment]
@visoft since you created this PR from your master branch, it may make sense to create a new feature branch starting from `ruby-docx:master` and include only the relevant commits.

File: lib/docx/document.rb [hunk 3]
```diff
@@ -20,9 +20,16 @@ module Docx
...
+    def initialize(path_or_io, options = {})
       @replace = {}
-      @zip = Zip::File.open(path)
+
+      # if path-or_io is string && does not contain a null byte
+      if (path_or_io.instance_of?(String) && !/\u0000/.match?(path_or_io))
+        @zip = Zip::File.open(path_or_io)
+      else
+        @zip = Zip::File.open_buffer(path_or_io)
+      end
```
Comment (satoryu) [inline lib/docx/document.rb]: LGTM

**Pull request merged**